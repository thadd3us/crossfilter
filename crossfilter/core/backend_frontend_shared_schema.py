"""Shared schema definitions for backend-frontend communication."""

from dataclasses import dataclass
from enum import StrEnum
from typing import Any, Optional

from pydantic import BaseModel, Field


class ProjectionType(StrEnum):
    """Types of data projections supported by crossfilter."""
    
    TEMPORAL = "temporal"
    GEO = "geo"
    CLIP_EMBEDDING = "clip_embedding"  # placeholder for future


class FilterOperatorType(StrEnum):
    """Types of filter operations supported by crossfilter."""
    
    INTERSECTION = "intersection"  # Limit to current selection
    SUBTRACTION = "subtraction"    # Remove current selection


@dataclass(frozen=True)
class FilterEvent:
    """Represents a filter event from a specific projection."""
    
    projection_type: ProjectionType
    selected_df_ids: set[int]
    filter_operator: FilterOperatorType


# API Request Models
class LoadDataRequest(BaseModel):
    """Request model for loading data."""
    
    file_path: str


class DfIdsFilterRequest(BaseModel):
    """Request model for filtering to specific df_ids from a plot."""
    
    df_ids: list[int] = Field(..., description="df_ids from the plot (could be from lasso selection, visible area, etc.)")
    event_source: ProjectionType = Field(..., description="Which plot type this filtering comes from")
    filter_operator: FilterOperatorType = Field(..., description="Type of filter operation to apply")


# API Response Models
class SessionStateResponse(BaseModel):
    """Response model for session state information."""
    
    all_rows_count: int
    filtered_rows_count: int
    columns: list[str]
    memory_usage_mb: str
    temporal_projection: dict[str, Any]
    geo_projection: dict[str, Any]
    # Frontend-compatible aliases
    has_data: bool
    row_count: int
    filtered_count: int


class LoadDataResponse(BaseModel):
    """Response model for data loading operations."""
    
    success: bool
    message: str
    session_state: SessionStateResponse


class FilterResponse(BaseModel):
    """Response model for filter operations."""
    
    success: bool
    filter_state: SessionStateResponse


class TemporalPlotResponse(BaseModel):
    """Response model for temporal plot data."""
    
    plotly_plot: dict[str, Any]
    data_type: str
    point_count: int
    distinct_point_count: int
    aggregation_level: Optional[str]


class GeoPlotResponse(BaseModel):
    """Response model for geo plot data."""
    
    plotly_plot: dict[str, Any]
    marker_count: int
    distinct_point_count: int
    aggregation_level: Optional[str]


# Server-Sent Events Models
class SSEEvent(BaseModel):
    """Base model for Server-Sent Events."""
    
    type: str
    timestamp: float
    version: int


class FilterChangeEvent(SSEEvent):
    """Server-Sent Event for filter changes."""
    
    affected_components: list[str]
    session_state: SessionStateResponse


class HeartbeatEvent(SSEEvent):
    """Server-Sent Event for heartbeat."""
    
    pass


class ConnectionEstablishedEvent(FilterChangeEvent):
    """Server-Sent Event for connection establishment."""
    
    pass


