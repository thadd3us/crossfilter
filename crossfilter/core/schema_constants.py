"""Schema constants and enums for crossfilter data structures."""

from enum import StrEnum


class SchemaColumns(StrEnum):
    """Column names from the DataSchema."""

    UUID_STRING = "UUID_STRING"
    DATA_TYPE = "DATA_TYPE"
    NAME = "NAME"
    CAPTION = "CAPTION"
    SOURCE_FILE = "SOURCE_FILE"
    TIMESTAMP_MAYBE_TIMEZONE_AWARE = "TIMESTAMP_MAYBE_TIMEZONE_AWARE"
    TIMESTAMP_UTC = "TIMESTAMP_UTC"
    GPS_LATITUDE = "GPS_LATITUDE"
    GPS_LONGITUDE = "GPS_LONGITUDE"
    RATING_1_TO_5 = "RATING_1_TO_5"
    SIZE_IN_BYTES = "SIZE_IN_BYTES"


class FilterOperationType(StrEnum):
    """Types of filter operations."""

    SPATIAL = "spatial"
    TEMPORAL = "temporal"
    RESET = "reset"


class TemporalLevel(StrEnum):
    """Temporal quantization levels."""

    SECOND = "SECOND"
    MINUTE = "MINUTE"
    HOUR = "HOUR"
    DAY = "DAY"
    MONTH = "MONTH"
    YEAR = "YEAR"




# Standard DataFrame ID column name
DF_ID_COLUMN = "df_id"


# Helper functions to get quantized column names
def get_h3_column_name(level: int) -> str:
    """Get the H3 column name for a specific level (0-15)."""
    if not 0 <= level <= 15:
        raise ValueError(f"H3 level must be between 0 and 15, got {level}")
    return f"QUANTIZED_H3_L{level}"


def get_temporal_column_name(level: TemporalLevel) -> str:
    """Get the temporal quantization column name for a specific level."""
    return f"QUANTIZED_TIMESTAMP_{level}"
