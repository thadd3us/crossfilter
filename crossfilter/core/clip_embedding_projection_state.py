"""CLIP embedding projection state management for crossfilter visualization."""

import logging
from typing import Optional

import pandas as pd

from crossfilter.core.bucketing import (
    bucket_by_target_column,
    get_optimal_semantic_embedding_umap_h3_level,
)
from crossfilter.core.schema import get_semantic_embedding_umap_h3_column_name
from crossfilter.core.projection_state import ProjectionState
from crossfilter.core.schema import SchemaColumns as C

logger = logging.getLogger(__name__)


class ClipEmbeddingProjectionState:
    """
    Manages the CLIP embedding projection state for visualizing semantic similarity in 2D space.

    This class handles the spatial aggregation of CLIP embedding data points using H3 hexagonal
    indexing on the UMAP-projected coordinates, maintaining both the current projection DataFrame
    and the aggregation granularity level. The projection may be aggregated at some H3 level (with
    a COUNT column) or may show individual points (no COUNT column).
    """

    def __init__(self, max_rows: int) -> None:
        """
        Initialize CLIP embedding projection state.

        Args:
            max_rows: Maximum number of rows to display before aggregation
        """
        self.projection_state = ProjectionState(max_rows)
        self.current_h3_level: Optional[int] = None

        # User selectable.
        self.max_marker_size: int = 20

    def update_projection(self, filtered_rows: pd.DataFrame) -> None:
        """
        Update the CLIP embedding projection based on the current filtered data.

        Args:
            filtered_rows: Current filtered subset of all_rows
        """
        # Check if semantic embedding UMAP columns are present in the data
        if (
            C.SEMANTIC_EMBEDDING_UMAP_LATITUDE not in filtered_rows.columns
            or C.SEMANTIC_EMBEDDING_UMAP_LONGITUDE not in filtered_rows.columns
        ):
            logger.info(
                "Semantic embedding UMAP columns not found in data, skipping semantic embedding projection update"
            )
            # Set empty projection with no buckets
            self.projection_state.current_bucketing_column = None
            self.current_h3_level = None
            self.projection_state.projection_df = pd.DataFrame()
            return

        # Drop rows with missing semantic embedding UMAP coordinates
        logger.info(
            f"Projecting {len(filtered_rows)=} rows with semantic embedding UMAP coordinates."
        )
        filtered_rows = filtered_rows.dropna(
            subset=[C.SEMANTIC_EMBEDDING_UMAP_LATITUDE, C.SEMANTIC_EMBEDDING_UMAP_LONGITUDE]
        )
        logger.info(f"Kept {len(filtered_rows)=} rows with semantic embedding UMAP coordinates")

        optimal_level = get_optimal_semantic_embedding_umap_h3_level(
            filtered_rows, self.projection_state.max_rows
        )

        if optimal_level is None:
            self.projection_state.current_bucketing_column = None
            self.current_h3_level = None
            self.projection_state.projection_df = filtered_rows
            return

        self.current_h3_level = optimal_level
        self.projection_state.current_bucketing_column = get_semantic_embedding_umap_h3_column_name(
            optimal_level
        )
        self.projection_state.projection_df = bucket_by_target_column(
            filtered_rows,
            self.projection_state.current_bucketing_column,
            self.projection_state.groupby_column,
        )
        logger.info(
            f"Bucketed CLIP embedding data at optimal H3 level {optimal_level=}, {len(self.projection_state.projection_df)=}"
        )

    def get_summary(self) -> dict:
        """Get a summary of the current CLIP embedding projection state."""
        summary = self.projection_state.get_summary()
        # Add CLIP embedding-specific information
        summary["h3_level"] = self.current_h3_level
        summary["target_column"] = self.projection_state.current_bucketing_column
        return summary
