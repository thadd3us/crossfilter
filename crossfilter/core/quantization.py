"""Data quantization utilities for spatial and temporal aggregation."""

import h3
import pandas as pd
import logging
from typing import Optional, List

from crossfilter.core.schema_constants import (
    SchemaColumns,
    QuantizedColumns,
    TemporalLevel,
    DF_ID_COLUMN,
    get_h3_column_name,
)

logger = logging.getLogger(__name__)

# H3 resolution levels to pre-compute (0-15, where higher = more granular)
H3_LEVELS = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]

# Temporal quantization levels
TEMPORAL_LEVELS = [
    TemporalLevel.SECOND,
    TemporalLevel.MINUTE,
    TemporalLevel.HOUR,
    TemporalLevel.DAY,
    TemporalLevel.MONTH,
    TemporalLevel.YEAR,
]


def add_quantized_columns(df: pd.DataFrame) -> pd.DataFrame:
    """
    Add all quantized columns to the DataFrame.

    Args:
        df: Input DataFrame

    Returns:
        DataFrame with added quantized columns
    """
    df = df.copy()

    # Add spatial quantization (H3 cells)
    if (
        SchemaColumns.GPS_LATITUDE in df.columns
        and SchemaColumns.GPS_LONGITUDE in df.columns
    ):
        df = _add_h3_columns(df)

    # Add temporal quantization
    if SchemaColumns.TIMESTAMP_UTC in df.columns:
        df = _add_temporal_columns(df)

    logger.info(f"Added quantized columns to DataFrame with {len(df)} rows")
    return df


def _add_h3_columns(df: pd.DataFrame) -> pd.DataFrame:
    """Add H3 hexagon columns at multiple resolutions."""
    for level in H3_LEVELS:
        col_name = get_h3_column_name(level)
        df[col_name] = df.apply(
            lambda row: (
                h3.latlng_to_cell(
                    row[SchemaColumns.GPS_LATITUDE],
                    row[SchemaColumns.GPS_LONGITUDE],
                    level,
                )
                if pd.notna(row[SchemaColumns.GPS_LATITUDE])
                and pd.notna(row[SchemaColumns.GPS_LONGITUDE])
                else None
            ),
            axis=1,
        )
    return df


def _add_temporal_columns(df: pd.DataFrame) -> pd.DataFrame:
    """Add temporal quantization columns at multiple levels."""
    # Convert to datetime if string
    timestamps = pd.to_datetime(df[SchemaColumns.TIMESTAMP_UTC])

    # Second level (round to nearest second)
    df[QuantizedColumns.QUANTIZED_TIMESTAMP_SECOND] = timestamps.dt.floor("s")

    # Minute level
    df[QuantizedColumns.QUANTIZED_TIMESTAMP_MINUTE] = timestamps.dt.floor("min")

    # Hour level
    df[QuantizedColumns.QUANTIZED_TIMESTAMP_HOUR] = timestamps.dt.floor("h")

    # Day level
    df[QuantizedColumns.QUANTIZED_TIMESTAMP_DAY] = timestamps.dt.floor("D")

    # Month level - normalize to first day of month
    df[QuantizedColumns.QUANTIZED_TIMESTAMP_MONTH] = (
        timestamps.dt.normalize().dt.to_period("M").dt.start_time
    )

    # Year level - normalize to first day of year
    df[QuantizedColumns.QUANTIZED_TIMESTAMP_YEAR] = (
        timestamps.dt.normalize().dt.to_period("Y").dt.start_time
    )

    return df


def get_optimal_h3_level(df: pd.DataFrame, max_groups: int) -> Optional[int]:
    """
    Find the H3 resolution level that gives closest to max_groups unique cells.

    Args:
        df: DataFrame with H3 columns
        max_groups: Maximum number of groups desired

    Returns:
        Optimal H3 level, or None if no H3 columns found
    """
    best_level = None
    best_count = 0

    for level in H3_LEVELS:
        col_name = get_h3_column_name(level)
        if col_name in df.columns:
            unique_count = df[col_name].nunique()
            if unique_count <= max_groups and unique_count > best_count:
                best_level = level
                best_count = unique_count

    logger.debug(f"Optimal H3 level: {best_level} with {best_count} groups")
    return best_level


def get_optimal_temporal_level(
    df: pd.DataFrame, max_groups: int
) -> Optional[TemporalLevel]:
    """
    Find the temporal quantization level that gives closest to max_groups unique buckets.

    Args:
        df: DataFrame with temporal quantization columns
        max_groups: Maximum number of groups desired

    Returns:
        Optimal temporal level, or None if no temporal columns found
    """
    best_level = None
    best_count = 0

    for level in TEMPORAL_LEVELS:
        col_name = f"QUANTIZED_TIMESTAMP_{level.upper()}"
        if col_name in df.columns:
            unique_count = df[col_name].nunique()
            if unique_count <= max_groups and unique_count > best_count:
                best_level = level
                best_count = unique_count

    logger.debug(f"Optimal temporal level: {best_level} with {best_count} groups")
    return best_level


def aggregate_by_h3(df: pd.DataFrame, h3_level: int) -> pd.DataFrame:
    """
    Aggregate data by H3 cells at the specified level.

    Args:
        df: DataFrame with H3 columns
        h3_level: H3 resolution level to aggregate by

    Returns:
        Aggregated DataFrame with H3 cell, count, and representative lat/lon
    """
    col_name = get_h3_column_name(h3_level)
    if col_name not in df.columns:
        raise ValueError(f"H3 level {h3_level} not found in DataFrame")

    # Group by H3 cell and aggregate
    grouped = (
        df.groupby(col_name)
        .agg(
            {
                SchemaColumns.GPS_LATITUDE: ["mean", "count"],
                SchemaColumns.GPS_LONGITUDE: "mean",
            }
        )
        .reset_index()
    )

    # Flatten column names
    grouped.columns = [col_name, "lat", "count", "lon"]

    # Add df_ids for the aggregated groups
    df_id_groups = (
        df.groupby(col_name)
        .apply(lambda x: list(x.index), include_groups=False)
        .reset_index(name="df_ids")
    )
    grouped = grouped.merge(df_id_groups, on=col_name)

    logger.debug(
        f"Aggregated {len(df)} rows into {len(grouped)} H3 groups at level {h3_level}"
    )
    return grouped


def aggregate_by_temporal(
    df: pd.DataFrame, temporal_level: TemporalLevel
) -> pd.DataFrame:
    """
    Aggregate data by temporal buckets at the specified level.

    Args:
        df: DataFrame with temporal quantization columns
        temporal_level: Temporal level to aggregate by

    Returns:
        Aggregated DataFrame with timestamp bucket, count, and cumulative count
    """
    col_name = f"QUANTIZED_TIMESTAMP_{temporal_level.upper()}"
    if col_name not in df.columns:
        raise ValueError(f"Temporal level {temporal_level} not found in DataFrame")

    # Group by temporal bucket and aggregate
    grouped = df.groupby(col_name).size().reset_index(name="count")

    # Add df_ids for the aggregated groups
    df_id_groups = (
        df.groupby(col_name)
        .apply(lambda x: list(x.index), include_groups=False)
        .reset_index(name="df_ids")
    )
    grouped = grouped.merge(df_id_groups, on=col_name)

    # Sort by timestamp and add cumulative count for CDF
    grouped = grouped.sort_values(col_name)
    grouped["cumulative_count"] = grouped["count"].cumsum()

    logger.debug(
        f"Aggregated {len(df)} rows into {len(grouped)} temporal groups at level {temporal_level}"
    )
    return grouped
