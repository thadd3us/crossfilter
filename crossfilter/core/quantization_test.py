"""Tests for quantization module."""

import pytest
import pandas as pd
from datetime import datetime

from crossfilter.core.quantization import (
    add_quantized_columns,
    get_optimal_h3_level,
    get_optimal_temporal_level, 
    aggregate_by_h3,
    aggregate_by_temporal,
    H3_LEVELS,
    TEMPORAL_LEVELS
)
from crossfilter.core.schema_constants import (
    SchemaColumns,
    TemporalLevel,
    DF_ID_COLUMN,
    get_h3_column_name,
    get_temporal_column_name
)


@pytest.fixture
def sample_df():
    """Create a sample DataFrame for testing."""
    df = pd.DataFrame({
        SchemaColumns.UUID_STRING: [f"uuid_{i}" for i in range(10)],
        SchemaColumns.GPS_LATITUDE: [37.7749 + i * 0.01 for i in range(10)],
        SchemaColumns.GPS_LONGITUDE: [-122.4194 + i * 0.01 for i in range(10)],
        SchemaColumns.TIMESTAMP_UTC: [
            f"2024-01-01T{10 + i}:00:00Z" for i in range(10)
        ]
    })
    # Convert timestamp to datetime
    df[SchemaColumns.TIMESTAMP_UTC] = pd.to_datetime(df[SchemaColumns.TIMESTAMP_UTC], utc=True)
    # Set stable df_id index
    df.index.name = DF_ID_COLUMN
    return df


def test_add_quantized_columns_spatial(sample_df, snapshot) -> None:
    """Test adding spatial quantization columns."""
    result = add_quantized_columns(sample_df)
    
    # Should have original columns plus quantized H3 columns
    assert len(result.columns) > len(sample_df.columns)
    
    # Check that H3 columns were added and their contents
    h3_columns = {col: result[col].tolist() for col in result.columns 
                  if col.startswith('QUANTIZED_H3_L')}
    assert h3_columns == snapshot


def test_add_quantized_columns_temporal(sample_df, snapshot) -> None:
    """Test adding temporal quantization columns."""
    result = add_quantized_columns(sample_df)
    
    # Check that temporal columns were added and their contents
    temporal_columns = {
        get_temporal_column_name(TemporalLevel.SECOND): result[get_temporal_column_name(TemporalLevel.SECOND)].dt.strftime('%Y-%m-%d %H:%M:%S').tolist(),
        get_temporal_column_name(TemporalLevel.MINUTE): result[get_temporal_column_name(TemporalLevel.MINUTE)].dt.strftime('%Y-%m-%d %H:%M').tolist(),
        get_temporal_column_name(TemporalLevel.HOUR): result[get_temporal_column_name(TemporalLevel.HOUR)].dt.strftime('%Y-%m-%d %H:00').tolist(),
        get_temporal_column_name(TemporalLevel.DAY): result[get_temporal_column_name(TemporalLevel.DAY)].dt.strftime('%Y-%m-%d').tolist(),
        get_temporal_column_name(TemporalLevel.MONTH): result[get_temporal_column_name(TemporalLevel.MONTH)].dt.strftime('%Y-%m').tolist(),
        get_temporal_column_name(TemporalLevel.YEAR): result[get_temporal_column_name(TemporalLevel.YEAR)].dt.strftime('%Y').tolist()
    }
    assert temporal_columns == snapshot


def test_add_quantized_columns_missing_spatial(snapshot) -> None:
    """Test quantization when spatial columns are missing."""
    df = pd.DataFrame({
        SchemaColumns.UUID_STRING: ["uuid_1"],
        SchemaColumns.TIMESTAMP_UTC: [pd.Timestamp("2024-01-01T10:00:00Z", tz='UTC')]
    })
    df.index.name = DF_ID_COLUMN
    
    result = add_quantized_columns(df)
    
    # Check column presence and content
    column_info = {
        'present_columns': sorted(result.columns.tolist()),
        'temporal_hour_value': result[get_temporal_column_name(TemporalLevel.HOUR)].dt.strftime('%Y-%m-%d %H:00').iloc[0],
        'h3_columns_present': [col for col in result.columns if col.startswith('QUANTIZED_H3_L')]
    }
    assert column_info == snapshot


def test_add_quantized_columns_missing_temporal(snapshot) -> None:
    """Test quantization when temporal columns are missing."""
    df = pd.DataFrame({
        SchemaColumns.UUID_STRING: ["uuid_1"],
        SchemaColumns.GPS_LATITUDE: [37.7749],
        SchemaColumns.GPS_LONGITUDE: [-122.4194]
    })
    df.index.name = DF_ID_COLUMN
    
    result = add_quantized_columns(df)
    
    # Check column presence and content
    column_info = {
        'present_columns': sorted(result.columns.tolist()),
        'h3_columns_present': [col for col in result.columns if col.startswith('QUANTIZED_H3_L')],
        'temporal_columns_present': [col for col in result.columns if col.startswith('QUANTIZED_TIMESTAMP')],
        'h3_level_4_value': result[get_h3_column_name(H3_LEVELS[0])].iloc[0]
    }
    assert column_info == snapshot


def test_get_optimal_h3_level(sample_df) -> None:
    """Test finding optimal H3 level."""
    quantized = add_quantized_columns(sample_df)
    
    # With small dataset, should find a high-resolution level
    optimal = get_optimal_h3_level(quantized, max_groups=1000)
    assert optimal is not None
    assert optimal in H3_LEVELS
    
    # With very small max_groups, should find lower resolution
    optimal_small = get_optimal_h3_level(quantized, max_groups=1)
    assert optimal_small is not None
    assert optimal_small <= optimal  # Lower resolution has smaller numbers


def test_get_optimal_temporal_level(sample_df) -> None:
    """Test finding optimal temporal level."""
    quantized = add_quantized_columns(sample_df)
    
    # With small dataset, should find a fine-grained level
    optimal = get_optimal_temporal_level(quantized, max_groups=1000)
    assert optimal is not None
    assert optimal in TEMPORAL_LEVELS
    
    # With very small max_groups, should find coarser level
    optimal_small = get_optimal_temporal_level(quantized, max_groups=1)
    assert optimal_small is not None


def test_aggregate_by_h3(sample_df) -> None:
    """Test H3 aggregation."""
    quantized = add_quantized_columns(sample_df)
    h3_level = 7  # Use a specific level for testing
    
    aggregated = aggregate_by_h3(quantized, h3_level)
    
    # Should have aggregated columns
    expected_cols = [
        get_h3_column_name(h3_level),
        'lat', 'count', 'lon', 'df_ids'
    ]
    for col in expected_cols:
        assert col in aggregated.columns
    
    # Count should be sum of original rows
    assert aggregated['count'].sum() == len(sample_df)
    
    # df_ids should contain lists of original df_ids
    assert all(isinstance(df_ids, list) for df_ids in aggregated['df_ids'])
    
    # All original df_ids should be present
    all_df_ids = []
    for df_ids in aggregated['df_ids']:
        all_df_ids.extend(df_ids)
    assert set(all_df_ids) == set(sample_df.index)


def test_aggregate_by_temporal(sample_df) -> None:
    """Test temporal aggregation."""
    quantized = add_quantized_columns(sample_df)
    temporal_level = TemporalLevel.HOUR
    
    aggregated = aggregate_by_temporal(quantized, temporal_level)
    
    # Should have aggregated columns
    expected_cols = [
        f"QUANTIZED_TIMESTAMP_{temporal_level.upper()}",
        'count', 'df_ids', 'cumulative_count'
    ]
    for col in expected_cols:
        assert col in aggregated.columns
    
    # Count should be sum of original rows
    assert aggregated['count'].sum() == len(sample_df)
    
    # Cumulative count should end at total count
    assert aggregated['cumulative_count'].iloc[-1] == len(sample_df)
    
    # Should be sorted by timestamp
    timestamp_col = f"QUANTIZED_TIMESTAMP_{temporal_level.upper()}"
    assert aggregated[timestamp_col].is_monotonic_increasing


def test_aggregate_by_h3_invalid_level(sample_df) -> None:
    """Test H3 aggregation with invalid level."""
    quantized = add_quantized_columns(sample_df)
    
    with pytest.raises(ValueError, match="H3 level must be between 0 and 15, got 99"):
        aggregate_by_h3(quantized, 99)


def test_aggregate_by_temporal_invalid_level(sample_df) -> None:
    """Test temporal aggregation with invalid level."""
    quantized = add_quantized_columns(sample_df)
    
    with pytest.raises(ValueError, match="Temporal level invalid not found"):
        aggregate_by_temporal(quantized, "invalid")


def test_empty_dataframe() -> None:
    """Test quantization with empty DataFrame."""
    df = pd.DataFrame(columns=[
        SchemaColumns.GPS_LATITUDE,
        SchemaColumns.GPS_LONGITUDE,
        SchemaColumns.TIMESTAMP_UTC
    ])
    df.index.name = DF_ID_COLUMN
    
    result = add_quantized_columns(df)
    
    # Should not raise errors
    assert len(result) == 0
    
    # Should have quantized columns
    assert get_h3_column_name(H3_LEVELS[0]) in result.columns


def test_null_coordinates() -> None:
    """Test quantization with null coordinates."""
    df = pd.DataFrame({
        SchemaColumns.UUID_STRING: ["uuid_1", "uuid_2"],
        SchemaColumns.GPS_LATITUDE: [37.7749, None],
        SchemaColumns.GPS_LONGITUDE: [-122.4194, None],
        SchemaColumns.TIMESTAMP_UTC: [
            pd.Timestamp("2024-01-01T10:00:00Z", tz='UTC'),
            pd.Timestamp("2024-01-01T11:00:00Z", tz='UTC')
        ]
    })
    df.index.name = DF_ID_COLUMN
    
    result = add_quantized_columns(df)
    
    # H3 columns should have null for null coordinates
    h3_col = get_h3_column_name(H3_LEVELS[0])
    assert pd.notna(result.loc[0, h3_col])  # Valid coordinates
    assert pd.isna(result.loc[1, h3_col])   # Null coordinates
    
    # Temporal columns should work for both rows
    assert pd.notna(result.loc[0, get_temporal_column_name(TemporalLevel.HOUR)])
    assert pd.notna(result.loc[1, get_temporal_column_name(TemporalLevel.HOUR)])