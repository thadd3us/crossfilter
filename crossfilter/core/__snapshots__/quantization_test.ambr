# serializer version: 1
# name: test_add_quantized_columns_missing_spatial
  dict({
    'h3_columns_present': list([
    ]),
    'present_columns': list([
      <QuantizedColumns.QUANTIZED_TIMESTAMP_DAY: 'QUANTIZED_TIMESTAMP_DAY'>,
      <QuantizedColumns.QUANTIZED_TIMESTAMP_HOUR: 'QUANTIZED_TIMESTAMP_HOUR'>,
      <QuantizedColumns.QUANTIZED_TIMESTAMP_MINUTE: 'QUANTIZED_TIMESTAMP_MINUTE'>,
      <QuantizedColumns.QUANTIZED_TIMESTAMP_MONTH: 'QUANTIZED_TIMESTAMP_MONTH'>,
      <QuantizedColumns.QUANTIZED_TIMESTAMP_SECOND: 'QUANTIZED_TIMESTAMP_SECOND'>,
      <QuantizedColumns.QUANTIZED_TIMESTAMP_YEAR: 'QUANTIZED_TIMESTAMP_YEAR'>,
      <SchemaColumns.TIMESTAMP_UTC: 'TIMESTAMP_UTC'>,
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>,
    ]),
    'temporal_hour_value': '2024-01-01 10:00',
  })
# ---
# name: test_add_quantized_columns_missing_temporal
  dict({
    'h3_columns_present': list([
      'QUANTIZED_H3_L3',
      'QUANTIZED_H3_L4',
      'QUANTIZED_H3_L5',
      'QUANTIZED_H3_L6',
      'QUANTIZED_H3_L7',
      'QUANTIZED_H3_L8',
      'QUANTIZED_H3_L9',
      'QUANTIZED_H3_L10',
      'QUANTIZED_H3_L11',
      'QUANTIZED_H3_L12',
    ]),
    'h3_level_4_value': '832830fffffffff',
    'present_columns': list([
      <SchemaColumns.GPS_LATITUDE: 'GPS_LATITUDE'>,
      <SchemaColumns.GPS_LONGITUDE: 'GPS_LONGITUDE'>,
      'QUANTIZED_H3_L10',
      'QUANTIZED_H3_L11',
      'QUANTIZED_H3_L12',
      'QUANTIZED_H3_L3',
      'QUANTIZED_H3_L4',
      'QUANTIZED_H3_L5',
      'QUANTIZED_H3_L6',
      'QUANTIZED_H3_L7',
      'QUANTIZED_H3_L8',
      'QUANTIZED_H3_L9',
      <SchemaColumns.UUID_STRING: 'UUID_STRING'>,
    ]),
    'temporal_columns_present': list([
    ]),
  })
# ---
# name: test_add_quantized_columns_spatial
  dict({
    'QUANTIZED_H3_L10': list([
      '8a283082800ffff',
      '8a283082aa6ffff',
      '8a283082a28ffff',
      '8a283080cce7fff',
      '8a283080e86ffff',
      '8a283080e08ffff',
      '8a283081dae7fff',
      '8a283081d64ffff',
      '8a2830819527fff',
      '8a283081b8c7fff',
    ]),
    'QUANTIZED_H3_L11': list([
      '8b283082800bfff',
      '8b283082aa6efff',
      '8b283082a288fff',
      '8b283080cce5fff',
      '8b283080e86afff',
      '8b283080e08bfff',
      '8b283081dae3fff',
      '8b283081d64cfff',
      '8b2830819524fff',
      '8b283081b8c5fff',
    ]),
    'QUANTIZED_H3_L12': list([
      '8c283082800b3ff',
      '8c283082aa6e1ff',
      '8c283082a288dff',
      '8c283080cce57ff',
      '8c283080e86a5ff',
      '8c283080e08bdff',
      '8c283081dae3bff',
      '8c283081d64c5ff',
      '8c28308195243ff',
      '8c283081b8c5bff',
    ]),
    'QUANTIZED_H3_L3': list([
      '832830fffffffff',
      '832830fffffffff',
      '832830fffffffff',
      '832830fffffffff',
      '832830fffffffff',
      '832830fffffffff',
      '832830fffffffff',
      '832830fffffffff',
      '832830fffffffff',
      '832830fffffffff',
    ]),
    'QUANTIZED_H3_L4': list([
      '8428309ffffffff',
      '8428309ffffffff',
      '8428309ffffffff',
      '8428309ffffffff',
      '8428309ffffffff',
      '8428309ffffffff',
      '8428309ffffffff',
      '8428309ffffffff',
      '8428309ffffffff',
      '8428309ffffffff',
    ]),
    'QUANTIZED_H3_L5': list([
      '85283083fffffff',
      '85283083fffffff',
      '85283083fffffff',
      '85283083fffffff',
      '85283083fffffff',
      '85283083fffffff',
      '85283083fffffff',
      '85283083fffffff',
      '85283083fffffff',
      '85283083fffffff',
    ]),
    'QUANTIZED_H3_L6': list([
      '86283082fffffff',
      '86283082fffffff',
      '86283082fffffff',
      '86283080fffffff',
      '86283080fffffff',
      '86283080fffffff',
      '86283080fffffff',
      '86283081fffffff',
      '86283081fffffff',
      '86283081fffffff',
    ]),
    'QUANTIZED_H3_L7': list([
      '872830828ffffff',
      '87283082affffff',
      '87283082affffff',
      '87283080cffffff',
      '87283080effffff',
      '87283080effffff',
      '87283081dffffff',
      '87283081dffffff',
      '872830819ffffff',
      '87283081bffffff',
    ]),
    'QUANTIZED_H3_L8': list([
      '8828308281fffff',
      '88283082abfffff',
      '88283082a3fffff',
      '88283080cdfffff',
      '88283080e9fffff',
      '88283080e1fffff',
      '88283081dbfffff',
      '88283081d7fffff',
      '882830819dfffff',
      '88283081b9fffff',
    ]),
    'QUANTIZED_H3_L9': list([
      '89283082803ffff',
      '89283082aa7ffff',
      '89283082a2bffff',
      '89283080ccfffff',
      '89283080e87ffff',
      '89283080e0bffff',
      '89283081dafffff',
      '89283081d67ffff',
      '89283081953ffff',
      '89283081b8fffff',
    ]),
  })
# ---
# name: test_add_quantized_columns_temporal
  dict({
    <QuantizedColumns.QUANTIZED_TIMESTAMP_DAY: 'QUANTIZED_TIMESTAMP_DAY'>: list([
      '2024-01-01',
      '2024-01-01',
      '2024-01-01',
      '2024-01-01',
      '2024-01-01',
      '2024-01-01',
      '2024-01-01',
      '2024-01-01',
      '2024-01-01',
      '2024-01-01',
    ]),
    <QuantizedColumns.QUANTIZED_TIMESTAMP_HOUR: 'QUANTIZED_TIMESTAMP_HOUR'>: list([
      '2024-01-01 10:00',
      '2024-01-01 11:00',
      '2024-01-01 12:00',
      '2024-01-01 13:00',
      '2024-01-01 14:00',
      '2024-01-01 15:00',
      '2024-01-01 16:00',
      '2024-01-01 17:00',
      '2024-01-01 18:00',
      '2024-01-01 19:00',
    ]),
    <QuantizedColumns.QUANTIZED_TIMESTAMP_MINUTE: 'QUANTIZED_TIMESTAMP_MINUTE'>: list([
      '2024-01-01 10:00',
      '2024-01-01 11:00',
      '2024-01-01 12:00',
      '2024-01-01 13:00',
      '2024-01-01 14:00',
      '2024-01-01 15:00',
      '2024-01-01 16:00',
      '2024-01-01 17:00',
      '2024-01-01 18:00',
      '2024-01-01 19:00',
    ]),
    <QuantizedColumns.QUANTIZED_TIMESTAMP_MONTH: 'QUANTIZED_TIMESTAMP_MONTH'>: list([
      '2024-01',
      '2024-01',
      '2024-01',
      '2024-01',
      '2024-01',
      '2024-01',
      '2024-01',
      '2024-01',
      '2024-01',
      '2024-01',
    ]),
    <QuantizedColumns.QUANTIZED_TIMESTAMP_SECOND: 'QUANTIZED_TIMESTAMP_SECOND'>: list([
      '2024-01-01 10:00:00',
      '2024-01-01 11:00:00',
      '2024-01-01 12:00:00',
      '2024-01-01 13:00:00',
      '2024-01-01 14:00:00',
      '2024-01-01 15:00:00',
      '2024-01-01 16:00:00',
      '2024-01-01 17:00:00',
      '2024-01-01 18:00:00',
      '2024-01-01 19:00:00',
    ]),
    <QuantizedColumns.QUANTIZED_TIMESTAMP_YEAR: 'QUANTIZED_TIMESTAMP_YEAR'>: list([
      '2024',
      '2024',
      '2024',
      '2024',
      '2024',
      '2024',
      '2024',
      '2024',
      '2024',
      '2024',
    ]),
  })
# ---
